back/app/services/invites/xlsx_processor.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+class Invites::XlsxProcessor
+  def initialize(error_storage)
+    @error_storage = error_storage
+  end
+
+  def param_to_hash_array(xlsx_param)
+    # Strip out data;...base64 prefix if it's there
+    start = xlsx_param.index(';base64,')
+    pure_base64 = start ? xlsx_param[(start + 8)..] : xlsx_param
+
+    xlsx_file = StringIO.new(Base64.decode64(pure_base64))
+
+    map_rows = []
+    old_row = 0
+    hash_array = XlsxService.new.xlsx_to_hash_array(xlsx_file).select do |invite_params|
+      if invite_params.present?
+        map_rows += [old_row]
+      end
+      old_row += 1
+      invite_params.present?
+    end
+
+    postprocess_xlsx_hash_array(hash_array)
+
+    [hash_array, map_rows]
+  end
+
+  private
+
+  def postprocess_xlsx_hash_array(hash_array)
+    hash_array.each.with_index do |hash, row_index|
+      @current_row = row_index
+      if hash['groups']
+        hash['group_ids'] = xlsx_groups_to_group_ids(hash['groups'])
+      end
+      hash.delete('groups')
+
+      if hash['admin'].present?
+        hash['roles'] = xlsx_admin_to_roles(hash['admin'])
+      end
+      hash.delete('admin')
+
+      if hash['language'].present?
+        hash['locale'] = hash['language']
+      end
+      hash.delete('language')
+
+      if hash['send_invite_email'].present?
+        hash['send_invite_email'] = Utils.to_bool(hash['send_invite_email'])
+      end
+
+      coerce_custom_field_types(hash)
+    end
+  rescue StandardError => e
+    @error_storage.add_error(:unparseable_excel, raw_error: e.to_s)
+    @error_storage.fail_now
+  ensure
+    @current_row = nil
+  end
+
+  def xlsx_groups_to_group_ids(groups)
+    groups.split(',').filter_map do |group_title|
+      stripped_group_title = group_title.strip
+      group = Group.all.find { |g| g.title_multiloc.values.map(&:strip).include? stripped_group_title }&.id
+      group || (@error_storage.add_error(:unknown_group, row: @current_row, value: stripped_group_title) && nil)
+    end
+  rescue StandardError => e
+    @error_storage.add_error(:malformed_groups_value, row: @current_row, value: groups, raw_error: e.to_s)
+    []
+  end
+
+  def xlsx_admin_to_roles(admin)
+    if [true, 'TRUE', 'true', '1', 1].include? admin
+      [{ 'type' => 'admin' }]
+    elsif [false, 'FALSE', 'false', '0', 0].include? admin
+      []
+    else
+      @error_storage.add_error(:malformed_admin_value, row: @current_row, value: admin)
+      []
```

Comment:
Can you make use of `Utils.to_bool` here?
```
case Utils.to_bool(admin)
when true
...
when false
...
else
...
```

Reply:
> so maybe that's why you didn't?

That was a part of my reasoning. But the main reason is I didn't want to change any code to be more sure I don't break anything.